src/schematools/utils.py

```diff
@@ -92,9 +106,17 @@ def def_from_url(base_url: str, data_type: Type[types.ST], dataset_id: str) -> t
         response = connection.get(f"{base_url}{index[dataset_id]}")
         response.raise_for_status()
 
-        schema_lookup[dataset_id] = data_type.from_dict(response.json())
+        dataset_schema = data_type.from_dict(response.json())
 
-    return schema_lookup[dataset_id]
+    # For this recursive call, we set prefetch_related=False
+    # to avoid deep/endless recursion
+    # The result of def_from_url does not need to be stored,
+    # because is it cached on the DatasetSchema instances.
+    if prefetch_related:
+        for ds_id in dataset_schema.related_dataset_schema_ids:
```

Comment:
We have a typing issue here. `data_type` can be any `SchemaType`. However only `DatasetSchema`'s have a `related_dataset_schema_ids` attribute. MyPy warns for this by stating line 109 needs an annotation. If we put that annotation there, namely

```python
dataset_schema: types.ST = data_type.from_dict(response.json())
```
then MyPy displays the error:  

```console
 mypy src/schematools/utils.py
src/schematools/utils.py:115:22: error: "types.ST" has no attribute "related_dataset_schema_ids"  [attr-defined]
```

So either the argument `data_type` needs to be limited to `Type[types.DatasetSchema]` or we need an explicit test to see if `dataset_schema` is indeed an `instanceof` `types.DatasetSchema` before we reference `related_dataset_schema_ids`. I'm note sure what the correct solution is this particular case.

I was wondering why the MyPy pre-commit hook didn't catch this, but now I see that we don't have the MyPy  pre-commit hook enabled for this project.

Reply:
Fixed in https://github.com/Amsterdam/schema-tools/pull/153/ I added an `instanceof` check.